src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs

```diff
@@ -0,0 +1,490 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+using System.Net.Http;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Bindings.Runtime;
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.Azure.WebJobs.Script.Description.PowerShell;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class PowerShellFunctionInvoker : FunctionInvokerBase
+    {
+        private readonly ScriptHost _host;
+        private readonly string _scriptFilePath;
+        private readonly string _functionName;
+
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly IMetricsLogger _metrics;
+
+        private string _script;
+        private List<string> _moduleFiles;
+        private Dictionary<string, string> _environmentVariables;
+        private Action _reloadScript;
+
+        private static readonly string[] WatchedFileTypes = { ".ps1", ".psm1", ".dll", ".psd1" };
+
+        internal PowerShellFunctionInvoker(ScriptHost host, FunctionMetadata functionMetadata,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings) : base(host, functionMetadata)
+        {
+            _host = host;
+            _scriptFilePath = functionMetadata.Source;
+            _functionName = functionMetadata.Name;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _metrics = host.ScriptConfig.HostConfig.GetService<IMetricsLogger>();
+            _environmentVariables = new Dictionary<string, string>();
+            InitializeFileWatcherIfEnabled();
+            _reloadScript = ReloadScript;
+            _reloadScript = _reloadScript.Debounce();
+        }
+
+        public override async Task Invoke(object[] parameters)
+        {
+            object input = parameters[0];
+            TraceWriter traceWriter = (TraceWriter)parameters[1];
+            IBinderEx binder = (IBinderEx)parameters[2];
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters[3];
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            FunctionStartedEvent startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                object convertedInput = input;
```

Comment:
It's a TODO on me to refactor some of this common code out so it is shared rather than duplicated across the various providers (PowerShell and Script).